### PR TITLE
feat(scalasdk): EventSourcedEntity Public API.

### PR DESCRIPTION
**Pull request by RayRoestenburg**
Public Scala API for EventSourcedEntity.

### CODE REVIEW

File: sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala [hunk 1]
```diff
@@ -63,70 +63,69 @@ class EventSourcedEntityEffectImpl[S] extends Builder[S] with OnSuccessBuilder[S
     secondary
   }
 
-  override def emitEvent(event: Any): OnSuccessBuilder[S] = {
+  override def emitEvent(event: Any): EventSourcedEntityEffectImpl[S] = {
```
Comment (johanandren) [inline sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: Hmm, isn't the whole point of working with a builder to have a two step thing where you first have to specify the primary effect, get a OnSuccessBuilder back which limits the possible additional effects to the secondary ones (`emitEvent(event).thenReply(state => some reply)`) or is that solved some way in the public API I'm missing?

**Review by johanandren — COMMENTED** [review]
Looks pretty straightforward 👍

File: sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala (hunk 1 above)
Comment (RayRoestenburg) [reply sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: I'm just not narrowing down the type so I can delegate to it.

**Review by octonato — COMMENTED** [review]
Looking good. Left a few comments.

File: sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala (hunk 1 above)
Comment (octonato) [reply sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: Yeah, I just realised that the original trait still forces the user to follow the builder. Though, I don't see why we needed to change it. We haven't done so for the others.

[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: This is the implementation

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/AkkaServerless.scala [hunk 2]
```diff
@@ -130,7 +130,7 @@ class AkkaServerless private (impl: javasdk.AkkaServerless) {
    *   This stateful service builder.
    */
   def register[S, E <: EventSourcedEntity[S]](provider: EventSourcedEntityProvider[S, E]): AkkaServerless =
-    AkkaServerless(impl.register(provider.impl))
+    AkkaServerless(impl.register(new JavaEventSourcedEntityProviderAdapter(provider)))
```
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/AkkaServerless.scala:133]: ok I'll do that no worries.

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/eventsourcedentity/EventSourcedEntityAdapters.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+private[scalasdk] final class JavaEventSourcedEntityProviderAdapter[S, E <: EventSourcedEntity[S]](
+    scalasdkProvider: EventSourcedEntityProvider[S, E])
+    extends javasdk.eventsourcedentity.EventSourcedEntityProvider[S, javasdk.eventsourcedentity.EventSourcedEntity[S]] {
+
+  def additionalDescriptors(): Array[Descriptors.FileDescriptor] = scalasdkProvider.additionalDescriptors.toArray
+  def entityType(): String = scalasdkProvider.entityType
+  def newHandler(context: javasdk.eventsourcedentity.EventSourcedEntityContext)
+      : javasdk.impl.eventsourcedentity.EventSourcedEntityHandler[
+        S,
+        javasdk.eventsourcedentity.EventSourcedEntity[S]] = {
```
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/eventsourcedentity/EventSourcedEntityAdapters.scala]: I changed this now..

File: sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala (hunk 1 above)
Comment (RayRoestenburg) [reply sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: Just to add @octonato and I talked about it, it's for some easy wrapping.

Comment (octonato) [reply sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntityEffectImpl.scala:66]: Yep, makes sense and users won't see it. Akka Projections does the same trick in a few places, but for other reasons.

**Review by octonato — APPROVED** [review]
LGTM

**Pull request merged**